ktor-server/ktor-server-plugins/ktor-server-di/common/src/io/ktor/server/plugins/di/DependencyResolution.kt

```diff
@@ -0,0 +1,95 @@
+ */
+
+package io.ktor.server.plugins.di
+
+import io.ktor.util.reflect.*
+import kotlin.getValue
+import kotlin.reflect.KProperty
+
+/**
+ * Functional interface for generating a fresh `DependencyResolver`.
+ */
+public fun interface DependencyResolution {
+    /**
+     * Resolves and creates a new instance of `DependencyResolver` using the provided `DependencyProvider`
+     * and `DependencyReflection`.
+     *
+     * @param provider The `DependencyProvider` instance responsible for managing dependency initializers
+     *                 and declarations.
+     * @param reflection The `DependencyReflection` instance used for reflective creation of dependency
+     *                   instances.
+     * @return A new instance of `DependencyResolver` configured with the provided arguments
+     */
+    public fun resolve(provider: DependencyProvider, reflection: DependencyReflection): DependencyResolver
+}
+
+/**
+ * A map of object instances.
+ */
+public interface DependencyMap {
+    public fun <T : Any> get(key: DependencyKey): T
+    public fun <T : Any> getOrPut(key: DependencyKey, defaultValue: () -> T): T
+}
+
+/**
+ * Extends [DependencyMap] with reflection, allowing for the automatic injection of types.
+ */
+public interface DependencyResolver : DependencyMap {
+    public val reflection: DependencyReflection
+}
+
+/**
+ * Basic implementation of [DependencyResolver] using a backing map.
+ *
+ * The map values are of the `Result` type so that we can ignore exceptions in the case of initialization problems for
+ * types that are never referenced.
+ */
+@Suppress("UNCHECKED_CAST")
+public class MapDependencyResolver(
+    override val reflection: DependencyReflection,
+    instances: Map<DependencyKey, Result<Any>>
+) : DependencyResolver {
+    private val map = instances.toMutableMap()
+
+    override fun <T : Any> get(key: DependencyKey): T =
+        (map[key] ?: throw MissingDependencyException(key)).getOrThrow() as T
+
+    override fun <T : Any> getOrPut(key: DependencyKey, defaultValue: () -> T): T =
+        map.getOrPut(key) { runCatching(defaultValue) }.getOrThrow() as T
+}
+
+/**
+ * Decorates the dependency resolver with a qualified name for the expected type.
+ *
+ * Useful with delegation when used like: `val connection by dependencies.named("postgres")`
+ */
+public fun DependencyResolver.named(key: String) =
+    DependencyResolverContext(this, key)
+
+/**
+ * Property delegation operator for resolving dependencies.
+ */
+public inline operator fun <reified T> DependencyMap.getValue(thisRef: Any?, property: KProperty<*>): T =
+    lazy<T> { resolve() }.getValue(thisRef, property)
+
+/**
+ * Property delegation for [DependencyResolverContext] for use with the `named` shorthand for string qualifiers.
+ */
+public inline operator fun <reified T> DependencyResolverContext.getValue(thisRef: Any?, property: KProperty<*>): T =
+    lazy<T> { resolver.resolve(key) }.getValue(thisRef, property)
```

Comment:
Lazy seems redundant as `getValue` is called immediately
```suggestion
public inline operator fun <reified T> DependencyMap.getValue(thisRef: Any?, property: KProperty<*>): T =
    resolve()

/**
 * Property delegation for [DependencyResolverContext] for use with the `named` shorthand for string qualifiers.
 */
public inline operator fun <reified T> DependencyResolverContext.getValue(thisRef: Any?, property: KProperty<*>): T =
    resolver.resolve(key)
```

Reply:
Yes, it seems my understanding of property delegation and lazy types was a bit off 😜